crypto/crypto.go

```diff
@@ -54,6 +61,47 @@ func Keccak256Hash(data ...[]byte) (h common.Hash) {
 	return h
 }
 
+// Keccak256Hasher creates a repetitive Keccak256 hasher, allowing the same hash
+// data structures to be reused between hash runs instead of requiring new ones
+// to be created.
+//
+// The returned function is not thread safe!
+func Keccak256Hasher() Hasher {
+	hasher := sha3.NewKeccak256()
+
+	return func(data []byte) []byte {
+		hasher.Write(data)
+		result := hasher.Sum(nil)
+		hasher.Reset()
+		return result
+	}
+}
+
+// Keccak512 calculates and returns the Keccak512 hash of the input data.
+func Keccak512(data ...[]byte) []byte {
+	d := sha3.NewKeccak512()
+	for _, b := range data {
+		d.Write(b)
+	}
+	return d.Sum(nil)
+}
+
+// Keccak512Hasher creates a repetitive Keccak512 hasher, allowing the same hash
+// data structures to be reused between hash runs instead of requiring new ones
+// to be created.
+//
+// The returned function is not thread safe!
+func Keccak512Hasher() Hasher {
+	hasher := sha3.NewKeccak512()
+
+	return func(data []byte) []byte {
+		hasher.Write(data)
+		result := hasher.Sum(nil)
+		hasher.Reset()
+		return result
+	}
+}
+
```

Comment:
This function is easy to implement on top of `hash.Hash`, there is no need to add specialized variants to crypto/sha3. 

```go
import "hash"

func NewHasher(h hash.Hash) func ([]byte) []byte {
    return func (data []byte) []byte {
        h.Write(data)
        result := h.Sum(nil)
        h.Reset()
        return result
    }
}